Instrument the dialog to measure startup time

shane-tomlinson opened the pull request:
A series of commits to start to instrument the dialog startup to find inefficiencies.

Commit:
Set the start_time based on the initial time the user clicked.

Commit:
Add startup related events to the KPI event stream.

Allows us to figure out where startup is taking a long time.

* window.opened
* window.javascript_start
* window.get_called
* xhr.* related events to see how long various XHR requests are taking.

Commit:
Adding the ability to store a duration on a KPI event.

Used in XHR requests initially to see how long each one takes from the user's perspective.

Commit:
Separating window.javascript_parse_start from window.javascript_parse_complete

seanmonstar commented on resources/static/common/js/browserid.js:
```diff
@@ -7,8 +7,7 @@
 
-  // Define some constants.
-  _.extend(window.BrowserID, {
+  var bid = BrowserID, constants = {
```
I don't see why the change is needed. Still, even if kept, the `bid` var is never referenced again.

shane-tomlinson replied:
I have moved this to be the very first script of the set, before any libraries. This is so I can get the exact time before any other scripts are parsed. Upon completion of script parsing, start.js is run. I can take the diff between parse end and parse begin to find the total parse time. If we start shaving script bytes and realize gains in parsing, I would like to be able to quantify how much time we are saving just in parsing.

seanmonstar replied:
I see, you're removing the need for underscore. However, when this is compressed into a single script, won't it have to parse the whole thing before it executes any of it?

shane-tomlinson replied:
I am unsure of how the browsers handle running code while parsing. javascript_parse may in fact not be the parse time, but the amount of time it takes to run all the statements in between the top of the script and the bottom. I need a better understanding of how browsers handle this.

At first glance, the NavigationTiming (https://dvcs.w3.org/hg/webperf/raw-file/tip/specs/NavigationTiming/Overview.html) API does not seem to offer any solutions for this either.

seanmonstar replied:
No, It doesn't. I assume it has to parse before executing, since it must do function hoisting.

A for-sure time would be a new `script` tag, above the main one, that only has `JS_START_TIME = new Date()` in it, and never gets combined into the rest.

seanmonstar replied:
However, that might be tainted by the download time of the main script.

Commit:
Combine "javascript_parse_start" and "javascript_parse_complete" to "javascript_parse"

Commit:
Remove the javascript_parse timer for now until a reliable way of measuring it can be found.

Commit:
Only measure dom_loading time if preffed on in configuration.

* Add new configuration option "measure_dom_loading" which defaults to false.
* Pass the value of measure_dom_loading to the dialog_layout template to indicate whether domLoading should be measured.
* In start.js, only measure dom_loading if measureDomLoading is set to true.

Commit:
Rename window.get_called to window.channel_established.

fmarier commented:
I left two small comments in the commits, but overall it looks good to me.

I guess the front-end tests that failed on Travis need to be fixed before merge.

Commit:
Add a note about measuring the dialog open time breaking CSP

Commit:
Change the start_time exception to be "new Error" based.

shane-tomlinson commented:
@fmarier - Added the note that enabling measuring the dom loading time breaks CSP.

The concern about the script in the HEAD not being wrapped I believe is because you were looking at an old version of the code.

And finally, fixed the unit test. This error was due to us changing all the exceptions to be "throw new Error" based instead of "throw <string>" based on dev, when I rebased I did not update this exception. This is now updated.

seanmonstar commented:
CSP doesn't allow inline scripts?

fmarier commented:
@seanmonstar If you want inline scripts or styles in CSP, you have to allow "unsafe-inline" which kinda defeats the purpose of CSP (protecting against an attacker's inline scripts).

fmarier commented:
Not sure why Travis is still marked as failed (maybe it hasn't run since Shane's last update?), but r+ from me.

Commit:
Fix the JSHint error in interaction_data.js
